### PR TITLE
[etcd] Add metric type mapping for metrics dataset

### PR DESCRIPTION
**Pull request by agithomas**
- Enhancement

## Proposed commit message

- Added metric_type mapping to leader datastream.
- To benefit from the TSDB enablement of the package.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

- elastic-package build
- elastic-package stack up -v -d --services package-registry

## Related issues
- https://github.com/elastic/integrations/issues/8327

### CODE REVIEW

**Comment by agithomas** [comment]
Mapping Details Added below

<details>
<summary> mapping details </summary>

```json

{
  "mappings": {
    "_meta": {
      "managed_by": "fleet",
      "managed": true,
      "package": {
        "name": "etcd"
      }
    },
    "_data_stream_timestamp": {
      "enabled": true
    },
    "dynamic_templates": [
      {
        "etcd.disk.wal_fsync_duration.ns.bucket.*": {
          "path_match": "etcd.disk.wal_fsync_duration.ns.bucket.*",
          "match_mapping_type": "long",
          "mapping": {
            "time_series_metric": "counter",
            "type": "long"
          }
        }
      },
      {
        "etcd.disk.backend_commit_duration.ns.bucket.*": {
          "path_match": "etcd.disk.backend_commit_duration.ns.bucket.*",
          "match_mapping_type": "long",
          "mapping": {
            "time_series_metric": "counter",
            "type": "long"
          }
        }
      },
      {
        "strings_as_keyword": {
          "match_mapping_type": "string",
          "mapping": {
            "ignore_above": 1024,
            "type": "keyword"
          }
        }
      }
    ],
    "date_detection": false,
    "properties": {
      "@timestamp": {
        "type": "date"
      },
      "agent": {
        "properties": {
          "ephemeral_id": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "id": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "name": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "type": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "version": {
            "type": "keyword",
            "ignore_above": 1024
          }
        }
      },
      "cloud": {
        "properties": {
          "account": {
            "properties": {
              "id": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          },
          "availability_zone": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "instance": {
            "properties": {
              "id": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "name": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          },
          "machine": {
            "properties": {
              "type": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          },
          "project": {
            "properties": {
              "id": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          },
          "provider": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "region": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "service": {
            "properties": {
              "name": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          }
        }
      },
      "data_stream": {
        "properties": {
          "dataset": {
            "type": "constant_keyword",
            "value": "etcd.metrics"
          },
          "namespace": {
            "type": "constant_keyword",
            "value": "default"
          },
          "type": {
            "type": "constant_keyword",
            "value": "metrics"
          }
        }
      },
      "ecs": {
        "properties": {
          "version": {
            "type": "keyword",
            "ignore_above": 1024
          }
        }
      },
      "elastic_agent": {
        "properties": {
          "id": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "snapshot": {
            "type": "boolean"
          },
          "version": {
            "type": "keyword",
            "ignore_above": 1024
          }
        }
      },
      "etcd": {
        "properties": {
          "api_version": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "disk": {
            "properties": {
              "backend_commit_duration": {
                "properties": {
                  "ns": {
                    "properties": {
                      "bucket": {
                        "properties": {
                          "1000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "2000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "4000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "8000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "16000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "32000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "64000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "128000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "256000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "512000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "1024000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "2048000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "4096000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "+Inf": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "8192000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          }
                        }
                      },
                      "count": {
                        "type": "long",
                        "time_series_metric": "counter"
                      },
                      "sum": {
                        "type": "long",
                        "time_series_metric": "counter"
                      }
                    }
                  }
                }
              },
              "mvcc_db_total_size": {
                "properties": {
                  "bytes": {
                    "type": "long",
                    "time_series_metric": "gauge"
                  }
                }
              },
              "wal_fsync_duration": {
                "properties": {
                  "ns": {
                    "properties": {
                      "bucket": {
                        "properties": {
                          "1000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "2000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "4000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "8000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "16000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "32000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "64000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "128000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "256000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "512000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "1024000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "2048000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "4096000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "+Inf": {
                            "type": "long",
                            "time_series_metric": "counter"
                          },
                          "8192000000": {
                            "type": "long",
                            "time_series_metric": "counter"
                          }
                        }
                      },
                      "count": {
                        "type": "long",
                        "time_series_metric": "counter"
                      },
                      "sum": {
                        "type": "long",
                        "time_series_metric": "counter"
                      }
                    }
                  }
                }
              }
            }
          },
          "memory": {
            "properties": {
              "go_memstats_alloc": {
                "properties": {
                  "bytes": {
                    "type": "long",
                    "time_series_metric": "gauge"
                  }
                }
              }
            }
          },
          "network": {
            "properties": {
              "client_grpc_received": {
                "properties": {
                  "bytes": {
                    "type": "long",
                    "time_series_metric": "counter"
                  }
                }
              },
              "client_grpc_sent": {
                "properties": {
                  "bytes": {
                    "type": "long",
                    "time_series_metric": "counter"
                  }
                }
              }
            }
          },
          "server": {
            "properties": {
              "grpc_handled": {
                "properties": {
                  "count": {
                    "type": "long",
                    "time_series_metric": "counter"
                  }
                }
              },
              "grpc_started": {
                "properties": {
                  "count": {
                    "type": "long",
                    "time_series_metric": "counter"
                  }
                }
              },
              "has_leader": {
                "type": "byte",
                "time_series_metric": "gauge"
              },
              "leader_changes": {
                "properties": {
                  "count": {
                    "type": "long",
                    "time_series_metric": "counter"
                  }
                }
              },
              "proposals_committed": {
                "properties": {
                  "count": {
                    "type": "long",
                    "time_series_metric": "gauge"
                  }
                }
              },
              "proposals_failed": {
                "properties": {
                  "count": {
                    "type": "long",
                    "time_series_metric": "counter"
                  }
                }
              },
              "proposals_pending": {
                "properties": {
                  "count": {
                    "type": "long",
                    "time_series_metric": "gauge"
                  }
                }
              }
            }
          }
        }
      },
      "event": {
        "properties": {
          "agent_id_status": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "dataset": {
            "type": "constant_keyword",
            "value": "etcd.metrics"
          },
          "duration": {
            "type": "long"
          },
          "ingested": {
            "type": "date",
            "format": "strict_date_time_no_millis||strict_date_optional_time||epoch_millis"
          },
          "module": {
            "type": "constant_keyword",
            "value": "etcd"
          }
        }
      },
      "host": {
        "properties": {
          "architecture": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "containerized": {
            "type": "boolean"
          },
          "hostname": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "id": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "ip": {
            "type": "ip"
          },
          "mac": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "name": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "os": {
            "properties": {
              "codename": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "family": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "kernel": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "name": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "platform": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "type": {
                "type": "keyword",
                "ignore_above": 1024
              },
              "version": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          }
        }
      },
      "metricset": {
        "properties": {
          "name": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "period": {
            "type": "long"
          }
        }
      },
      "service": {
        "properties": {
          "address": {
            "type": "keyword",
            "time_series_dimension": true
          },
          "type": {
            "type": "keyword",
            "ignore_above": 1024
          }
        }
      }
    }
  }
}

```
</details>

**Comment by agithomas** [comment]
Prometheus metrics API extract 

```
# TYPE etcd_server_has_leader gauge
# TYPE etcd_server_leader_changes_seen_total counter
# TYPE etcd_server_proposals_committed_total gauge
# TYPE etcd_server_proposals_pending gauge
# TYPE etcd_server_proposals_failed_total counter
# TYPE grpc_server_started_total counter
# TYPE grpc_server_handled_total counter
# TYPE etcd_mvcc_db_total_size_in_bytes gauge
# TYPE etcd_disk_wal_fsync_duration_seconds histogram
# TYPE etcd_disk_backend_commit_duration_seconds histogram

# TYPE go_memstats_alloc_bytes gauge
# TYPE etcd_network_client_grpc_sent_bytes_total counter
# TYPE etcd_network_client_grpc_received_bytes_total counter


```

**Review by ritalwar — APPROVED** [review]
LGTM!

File: packages/etcd/data_stream/metrics/fields/fields.yml [hunk 1]
```diff
@@ -6,30 +6,37 @@
       fields:
         - name: mvcc_db_total_size.bytes
           type: long
+          metric_type: gauge
           description: |
             Size of stored data at MVCC
         - name: wal_fsync_duration.ns.bucket.*
           type: long
+          metric_type: counter
```
Comment (lalit-satapathy) [inline packages/etcd/data_stream/metrics/fields/fields.yml:14]: @agithomas,

There are metrics named as` .sum` and `.count` which are mapped to counters (which seems correct based on metric names). 

There are couple of metrics (`_duration.ns.bucket.*`) which are also being mapped as counters. Can we cross-check, if these are counters?

Comment (agithomas) [reply packages/etcd/data_stream/metrics/fields/fields.yml:14]: @lalit-satapathy , please refer [here](https://github.com/elastic/integrations/pull/8351#issuecomment-1787180128) to know about the structure of mappings. 

Based on the existing mappings, it falls under `counter` metric_type and i have validated them against time

**Comment by lalit-satapathy** [comment]
It appears, this PR work overlaps with the work in Prometheus migration [PR](https://github.com/elastic/integrations/pull/8438/). In such case, we should clarify which one will be used going forward. 

CC: @gpop63

**Comment by mrodm** [comment]
Hi @agithomas, please update your branch with the latest contents from main branch. There was an important PR merged updating the CI pipelines. Thanks!

**Pull request closed without merging**